data_structures/linked_list/circular_doubly_linked_list.c

```diff
@@ -0,0 +1,230 @@
+        temp = (*start)->prev;
+        n1->next = *start;
+        (*start)->prev = n1;
+        n1->prev = temp;
+        temp->next = n1;
+        *start = n1;
+    }
+    printf("\nNode succesfully inserted\n\n");
+}
+
+/**
+ * @brief   Insert the node at end of list
+ * @param   start   starting pointer of list
+ * @param   data    the data which we want to insert into list
+ * @returns void
+ */
+void insert_at_last(struct node **start, int data)
+{
+    struct node *n1;
+    n1 = (struct node *)malloc(sizeof(struct node));
+    n1->value = data;
+    if (*start == NULL)
+    {
+        n1->next = n1;
+        n1->prev = n1;
+        *start = n1;
+    }
+    else
+    {
+        struct node *temp1, *temp2;
+        temp1 = *start;
+        temp2 = (*start)->prev;
+        n1->prev = temp2;
+        n1->next = temp1;
+        temp1->prev = n1;
+        temp2->next = n1;
+    }
+    printf("\nNode succesfully inserted\n\n");
+}
+
+/**
+ * @brief   Deletion of the first node in list function
+ * @param   start   starting pointer of list
+ * @returns void
+ */
+void delete_fist_node(struct node **start)
+{
+    if (*start == NULL)
+    {
+        printf("⚠️list is Empty⚠️\n");
+    }
+    else
+    {
+        struct node *temp1, *temp2;
+        temp1 = *start;
+        temp2 = temp1->prev;
+        if (temp1 == temp2)
+        {
+            free(temp2);
+            *start = NULL;
+        }
+        else
+        {
+            temp2->next = temp1->next;
+            (temp1->next)->prev = temp2;
+            *start = temp1->next;
+            free(temp1);
+            temp1 = NULL;
+        }
+        printf("\nNode succesfully deleted\n\n");
+    }
+}
+
+/**
+ * @brief   Deletion of the last node in list function
+ * @param   start   starting pointer of list
+ * @returns void
+ */
+void delete_last_node(struct node **start)
+{
+    if (*start == NULL)
+    {
+        printf("list is Empty\n");
+    }
+    else
+    {
+        struct node *temp1, *temp2;
+        temp1 = *start;
+        temp2 = temp1->prev;
+        if (temp1 == temp2)
+        {
+            free(temp2);
+            *start = NULL;
+        }
+        else
+        {
+            (temp2->prev)->next = temp1;
+            temp1->prev = temp2->prev;
+            free(temp2);
+            temp2 = NULL;
+        }
+        printf("\nNode succesfully deleted\n\n");
+    }
+}
+
+/**
+ * @brief   Display list function
+ * @param   start   starting pointer of list
+ * @returns void
+ */
+void Display(struct node *start)
+{
+    printf("\nIts Your Linked-list\n\n");
+    struct node *temp;
+    temp = start;
+    if (start != NULL)
+    {
+        while (temp->next != start)
+        {
+            printf("%lld -> ", temp->value);
+            temp = temp->next;
+        }
+        if (temp->next == start)
+        {
+            printf("%lld", temp->value);
+        }
+    }
+    else
+    {
+        printf("list is Empty\n");
+    }
+    printf("\n");
+}
+
```

Comment:
Hmm. Can we add some tests to make sure the functions work as excepted? We Insert a few instances at the start/last indexes, and we can do some small tests such as making sure it's sorted by running a sorting function, and making sure it matches certain numbers. What do you think?

Reply:
So you're suggesting that we'll make a list, write a function to sort it, and then assess if the results are accurate or not? Is it so?

Reply:
We can do that or make sure the insert/delete functions work fine only. Your choice. 🙂

Reply:
Sure, I will inform you and send the code once I am done with the implementation part. 👍🏻

Reply:
Awesome! Do let us know if you need any help!

Reply:
How are you going with the tests?